### PR TITLE
Epic 7.7.1 Syllabus Generator (AI Avengers) 

### PR DESCRIPTION
**Pull request by skreddypalvai**
# Pull Request - Syllabus Generator

## Summary:
This update introduces a new feature, 'Syllabus Generator' enabling educators to effortlessly create customized syllabus. By specifying the grade level, subject and course duration(number of weeks), users (educators) can generate a detailed and comprehensive syllabus. The generated syllabus includes the following sections:
* Course Title
* Course Description
* Course Objectives
* Required Materials
* Grading Policy
* Class Policy
* Course Outline( weekly)

This feature addresses the need for a simplified, yet comprehensive, method to develop syllabi tailored to specific educational requirements.

## Changes:
1. Added New Subfolder:
     * Created a ```syllabus_generator``` subfolder in the feature folder and added the following files:
       *  <kbd>tool.py</kbd>:  Contains the main logic for the syllabus generator tool, including Pydantic data structures and respective classes.
       * <kbd>core.py</kbd>: Includes the executor function for generating the syllabus.
       * <kbd>metadata.json</kbd>: Metadata information for feature API payload.
       * ``` prompt```  folder: Created a new folder and added <kbd>syllabi_gen.txt</kbd>, a prompt template for generating the syllabus based on user inputs.
       * ``` tests```  folder: Created a new folder and added <kbd>test_syllabi</kbd> test cases for the tool and core files to ensure functionality.
2.  Configuration Update:
    *  Updated <kbd>tools_config.json</kbd> file in the ```api```  folder to align with the new feature, ensuring proper integration and configuration.
 
Note: No deletions of the existing code were made.

##  Testing:
   *  Utilized `pytest` to test <kbd>tool.py</kbd> and <kbd>core.py</kbd>. Test cases covered the following scenarios:
      * Initialization of the 'SyllabusGenerator' class.
      * Compilation of the 'generator'.
      * Generation of the syllabus.
      * Execution of the 'executor' with valid input.
      * Execution of the 'executor' with invalid input.
      * Generation of the syllabus with invalid input.
      * Execution of the 'executor'  with invalid date format.
 
##  Results:
*   All unit tests passed, confirming the functionality of each component and handling of edge cases.

test cases

## Notes:
*  Educators have the option to add supplementary details such as start date, objectives, grading policy, class policy, and course outline.
* The main goal behind this enhancement is to streamline the syllabus creation process, making it more efficient and user friendly for educators.

## Screenshots:
* ### Sample Request:
 [sample request - syllabi.json](https://github.com/user-attachments/files/16452541/sample.request.-.syllabi.json)

INp Sylla

 
```
 {
  "user": {
    "id": "string",
    "fullName": "string",
    "email": "string"
  },
  "type": "tool",
  "tool_data": {
    "tool_id": 2,
    "inputs": [
      {
        "name": "grade_level",
        "value": "bachelor's"
      },
      {
        "name": "subject",
        "value": "computer science"
      },
      {
        "name": "num_weeks",
        "value": 15
      },
      {
        "name": "start_date",
        "value": "2024-08-01"
      },
      {
        "name": "additional_objectives",
        "value": ""
      },
      {
        "name": "additional_materials",
        "value": ""
      },
      {
        "name": "additional_grading_policy",
        "value": ""
      },
      {
        "name": "additional_class_policy",
        "value": ""
      },
      {
        "name": "custom_course_outline",
        "value": ""
      }
    ]
  }
} 
 ```

* ### Model output:
[response_1722491271875 - syllabi.json](https://github.com/user-attachments/files/16452569/response_1722491271875.-.syllabi.json)
Out Sylla 1
2  sylla
3  sylla

```
{
  "data": {
    "course_title": "Introduction to Computer Science",
    "grade_level": "bachelor's",
    "subject": "computer science",
    "num_weeks": 15,
    "start_date": "2024-08-01 00:00:00",
    "course_description": "This course provides a comprehensive introduction to the fundamental concepts of computer science, exploring key topics such as algorithms, data structures, programming languages, and software development principles. Students will gain practical experience through hands-on projects and assignments, developing essential problem-solving and computational thinking skills.",
    "course_objectives": [
      "Understand the role and importance of computer science in modern society.",
      "Develop proficiency in a high-level programming language.",
      "Analyze and solve problems using computational thinking principles.",
      "Design, implement, and test basic algorithms and data structures.",
      "Understand the software development lifecycle and apply it to small-scale projects."
    ],
    "additional_objectives": null,
    "required_materials": [
      "Introduction to Algorithms by Thomas H. Cormen",
      "Data Structures and Algorithms in Python by Michael T. Goodrich",
      "Code: The Hidden Language of Computer Hardware and Software by Charles Petzold",
      "Starting Out with Python by Tony Gaddis"
    ],
    "additional_materials": null,
    "grading_policy": {
      "assignments": 30,
      "quizzes": 15,
      "exams": 30,
      "participation": 5,
      "projects": 20,
      "other": 0
    },
    "additional_grading_policy": null,
    "class_policies": {
      "attendance": "While attendance is not mandatory, active participation and engagement in class discussions are highly encouraged.",
      "late_work": "Late submissions will be penalized 10% per day up to 3 days, after which no further submissions will be accepted.",
      "academic_integrity": "All work submitted must be original. Plagiarism or any form of academic dishonesty will not be tolerated.",
      "participation": "Active participation in class discussions and group activities is expected. This includes asking thoughtful questions, sharing insights, and contributing to a positive learning environment.",
      "special_considerations": "Students with disabilities who may require accommodations should contact the instructor at the beginning of the semester to discuss their needs."
    },
    "additional_class_policy": null,
    "course_outline": {
      "outline": [
        {
          "week": 1,
          "date": "2024-08-01",
          "topic": "Introduction to Computer Science and Problem Solving"
        },
        {
          "week": 2,
          "date": "2024-08-08",
          "topic": "Variables, Data Types, and Operators"
        },
        {
          "week": 3,
          "date": "2024-08-15",
          "topic": "Control Flow and Decision Making"
        },
        {
          "week": 4,
          "date": "2024-08-22",
          "topic": "Loops and Iteration"
        },
        {
          "week": 5,
          "date": "2024-08-29",
          "topic": "Functions and Modular Programming"
        },
        {
          "week": 6,
          "date": "2024-09-05",
          "topic": "Lists and Arrays"
        },
        {
          "week": 7,
          "date": "2024-09-12",
          "topic": "Strings and Text Processing"
        },
        {
          "week": 8,
          "date": "2024-09-19",
          "topic": "Dictionaries and Sets"
        },
        {
          "week": 9,
          "date": "2024-09-26",
          "topic": "File Input and Output"
        },
        {
          "week": 10,
          "date": "2024-10-03",
          "topic": "Introduction to Algorithms and Complexity"
        },
        {
          "week": 11,
          "date": "2024-10-10",
          "topic": "Searching and Sorting Algorithms"
        },
        {
          "week": 12,
          "date": "2024-10-17",
          "topic": "Object-Oriented Programming Concepts"
        },
        {
          "week": 13,
          "date": "2024-10-24",
          "topic": "Classes, Objects, and Inheritance"
        },
        {
          "week": 14,
          "date": "2024-10-31",
          "topic": "Software Development Principles"
        },
        {
          "week": 15,
          "date": "2024-11-07",
          "topic": "Final Project Presentations"
        }
      ]
    },
    "custom_course_outline": null
  }
}
```
## Test:
 * Clone the repo, create and activate a virtual environment and then install the necessary dependencies by running the following commands (Bash CLI):
     ```python -m venv venv```
    ```source venv/bin/activate ```
    ```pip install -r requirements.txt```
 *  Set Up Local Environment:
    * Configure your local google credentials key.
    * Set up your google API key.
    * Set the environment variable as 'dev'.
  *  Start the application by running ```./local-start.sh```. Refer to the JSON sample request provided above and test the future.

### CODE REVIEW

File: app/features/quizzify/core.py [hunk 1]
```diff
@@ -6,31 +6,31 @@
 
 logger = setup_logger()
 
-def executor(files: list[ToolFile], topic: str, num_questions: int, verbose=False):
-    
+
+def executor(files, topic: str, num_questions: int, verbose=False):
```
Comment (mikhailocampo) [inline app/features/quizzify/core.py:10]: It's good practice to have a type annotation here. Perhaps since it is no longer a "ToolFile" we should implement an abstract interface that the files references for type safety

File: app/features/quizzify/loaders.py
Comment (mikhailocampo) [inline app/features/quizzify/loaders.py:1]: Passing tests, I'll only skim since tests look OK

File: app/features/quizzify/tools.py
Comment (mikhailocampo) [inline app/features/quizzify/tools.py:1]: I'm sure these changes were meant to deal with the deprecated functions? Then should be good as long as this feature also didn't change

File: app/features/syllabus_generator/core.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+logger = setup_logger()
+
+def executor(grade_level: str, subject: str, num_weeks: int,
+             start_date: Optional[str] = None, additional_objectives: Optional[str] = None ,
+             additional_materials: Optional[str] = None, additional_grading_policy: Optional[str] = None,
+             additional_class_policy: Optional[str] = None, custom_course_outline: Optional[str] = None,verbose = True):
+    
+    # additional notes inlcluding start date are optional and Defaults to None
+    try:
+        if start_date:
```
Comment (mikhailocampo) [inline app/features/syllabus_generator/core.py:17]: I think this double try except block is a bit hard to read. Perhaps it can be refactored to be cleaner on catching the exceptions

File: app/features/syllabus_generator/prompt/syllabi_gen.txt
Comment (mikhailocampo) [inline app/features/syllabus_generator/prompt/syllabi_gen.txt:1]: Good prompts

File: app/features/syllabus_generator/tests/test_syllabi.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+def test_executor_handle_tool_executor_error(monkeypatch, setup_data):
+    def mock_generate(*args, **kwargs):
+        raise ToolExecutorError("Mock ToolExecutorError")
+
+    monkeypatch.setattr(SyllabusGenerator, "generate", mock_generate)
+
+    result = executor(**setup_data)
+    assert result.status == 500
+    assert "Mock ToolExecutorError" in result.message
```
Comment (mikhailocampo) [inline app/features/syllabus_generator/tests/test_syllabi.py:78]: You should use the `with pytest.exceptions` as its already built-in to handle expected exceptions to be raised

**Review by mikhailocampo — CHANGES_REQUESTED** [review]
Good work. See my changes requested. Also remove the `.DS_Store` throughout the commit, I think they're not needed and can go under `.gitignore`

File: app/features/quizzify/loaders.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+
+class CustomYoutubeLoader:
+
+    def __init__(self, verbose=None):
+        self.verbose = verbose
+
+    def transcript_check(self, video_id):
+        try:
+            transcript = YouTubeTranscriptApi.get_transcript(video_id=video_id,
```
Comment (AaronSosaRamos) [inline app/features/quizzify/loaders.py:171]: Use the LangChain document loader for a better implementation

File: app/features/quizzify/loaders.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+                                                             languages=['en', 'en-GB', 'en-US', 'en-AU', 'en-BZ',
```
Comment (AaronSosaRamos) [inline app/features/quizzify/loaders.py:172]: This could be a good point for multilang. support

File: app/features/quizzify/loaders.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            video_id = YoutubeLoader.extract_video_id(url)
+
+            data = ''
+
+            transcript_data = self.transcript_check(video_id)
+            if transcript_data:
+                try:
+                    
+                    # Filter transcript based on start_time and end_time
+                    if start_timestamp is not None and end_timestamp is not None:
```
Comment (AaronSosaRamos) [inline app/features/quizzify/loaders.py:198]: This is hardcoded. Please, be careful with these conditionals as they could be segregated using validation functions.

File: app/features/quizzify/tools.py [hunk 7]
```diff
@@ -183,21 +114,56 @@ def load(self, tool_files: List[ToolFile]) -> List[Document]:
...
 
+class BaseLoader:
+
+    def __init__(self, verbose = False):
+        self.verbose = verbose
+
+    def load(self, files):
+        
+        documents = []
+        file_type_list = ['csv','pptx','docx','pdf','txt']
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:134]: Please, decouple this for supporting more file types.

File: app/features/quizzify/tools.py [hunk 8]
```diff
@@ -183,21 +114,56 @@ def load(self, tool_files: List[ToolFile]) -> List[Document]:
...
+                other_files.append(file)
+            elif url.lower().startswith('http://') or url.lower().startswith('https://'):
+                webpage_files.append(file)
+            else:
+                raise ValueError(f"Received {file_type}, Unsupported File Type\n"
+                                    f"Supported File Types 'pdf, txt','web page url','pptx','csv','docx','Youtube Url'")
+
+
+        docs = CustomYoutubeLoader(verbose = self.verbose).load(youtube_files)
+        documents.extend(docs)
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:155]: This flow can be complex to accomplish if there is only one file type. Refactor this for a better workflow.

File: app/features/quizzify/tools.py [hunk 9]
```diff
@@ -208,97 +174,99 @@ def __init__(self, loader=None, splitter=None, vectorstore_class=None, embedding
...
-        pipeline = self.load_PDFs | self.split_loaded_documents | self.create_vectorstore
+
+        pipeline = self.load_data | self.split_loaded_documents | self.create_vectorstore
         return pipeline(documents)
 
+
 class QuizBuilder:
     def __init__(self, vectorstore, topic, prompt=None, model=None, parser=None, verbose=False):
         default_config = {
             "model": GoogleGenerativeAI(model="gemini-1.0-pro"),
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:236]: Change this model to `gemini-1.5-flash` or `gemini-1.5-pro`

File: app/features/quizzify/tools.py [hunk 10]
```diff
@@ -208,97 +174,99 @@ def __init__(self, loader=None, splitter=None, vectorstore_class=None, embedding
...
+
         chain = runner | prompt | self.model | self.parser
-        
+
         if self.verbose: logger.info(f"Chain compilation complete")
-        
+
         return chain
 
     def validate_response(self, response: Dict) -> bool:
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:272]: Is this required for the Syllabus Generator?

File: app/features/quizzify/tools.py [hunk 11]
```diff
@@ -320,15 +288,15 @@ def validate_response(self, response: Dict) -> bool:
 
     def format_choices(self, choices: Dict[str, str]) -> List[Dict[str, str]]:
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:289]: Is this required for the Syllabus Generator?

File: app/features/quizzify/tools.py [hunk 12]
```diff
@@ -320,15 +288,15 @@ def validate_response(self, response: Dict) -> bool:
...
         return [{"key": k, "value": v} for k, v in choices.items()]
-    
+
     def create_questions(self, num_questions: int = 5) -> List[Dict]:
```
Comment (AaronSosaRamos) [inline app/features/quizzify/tools.py:292]: Is this required for the Syllabus Generator?

File: app/features/syllabus_generator/core.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+            try:
+                datetime.strptime(start_date, "%Y-%m-%d")
+            except ValueError:
+                raise InputValidationError(f"Invalid date format for start_date: {start_date}")
+            
+        generator = SyllabusGenerator(grade_level, subject,num_weeks,start_date,
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/core.py:23]: Decouple this using a Pydantic Schema

File: app/features/syllabus_generator/sample request.json [hunk 14]
```diff
@@ -0,0 +1,49 @@
+{
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/sample request.json:1]: This sample request is not required for the main project. It must be in the README

File: app/features/syllabus_generator/tools.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        self.additional_objectives = additional_objectives or ""
+        self.additional_materials = additional_materials or ""
+        self.additional_grading_policy = additional_grading_policy or ""
+        self.additional_class_policy = additional_class_policy or ""
+        self.custom_course_outline = custom_course_outline or ""
+
+
+    def compile(self):
+        try:
+            input_variables = ["grade_level", "subject", "num_weeks"]
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:55]: What about the optional variables? How are they included in the main prompt?

File: app/features/syllabus_generator/tools.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+# data structure
+class GradingPolicy(BaseModel):
+    assignments: int = Field(description="Percentage of grade from assignments")
+    quizzes: int = Field(description="Percentage of grade from quizzes")
+    exams: int = Field(description="Percentage of grade from exams")
+    participation: int = Field(description="Percentage of grade from participation")
+    projects: int = Field(description="Percentage of grade from projects")
+    other: int = Field(description="Percentage of grade from other components")
+    
+    @validator('assignments', 'quizzes', 'exams', 'participation', 'projects', 'other')
```
Comment (AaronSosaRamos) [inline app/features/syllabus_generator/tools.py:113]: Good validation

**Review by AaronSosaRamos — CHANGES_REQUESTED** [review]
Please, solve the given comments

**Comment by AaronSosaRamos** [comment]
Closed due inactivity
